rfd/0078-trait-transforms.md

```diff
@@ -0,0 +1,754 @@
+
+If teleport admins cannot add or change the claims provided by their SSO
+provider, making simple rules like this can be extremely cumbersome or
+impossible within teleport.
+There are currently teleport users with deployments which automatically generate
+hundreds to thousands of roles to get around these limitations.
+
+## Details
+
+Example trait transform spec:
+
+```yaml
+kind: trait_transform
+version: v1
+metadata:
+  name: example
+spec:
+  # priority can be used to order the evaluation of multiple trait transforms in
+  # a cluster. Lower priorities will be evaluated first. Trait transforms with
+  # the same priority will be ordered by a lexicographical sort by their names. 
+  priority: 0
+
+  # filter is a predicate expression which will be applied to each incoming
+  # trait. `trait.key` and `trait.values` will be available to use in the
+  # expression. If the expression returns `false` for a given trait, it will be
+  # filtered out and excluded from further steps in the trait transform and from
+  # the user's certificate.
+  #
+  # In this example, only the "groups", "username", and "email" traits will be
+  # included. More powerful expressions are possible and will be explored later.
+  filter: >
+    contains(list("groups", "username", "email"), trait.key)
+
+  # override holds a map of trait keys to predicate expressions which should
+  # return the desired value for that trait. This can be used to override
+  # existing traits, or add new ones.
+  #
+  # All incoming traits are available for use in the expression by
+  # `external.<trait_name>` or `external["<trait_name>"]`.
+  #
+  # Various helper functions are provided, which will be explained later in the
+  # RFD.
+  override:
+    # groups will override the existing groups trait by appending the "dbs"
+    # group if the user's current groups includes "splunk".
+    groups: >
+      match(contains(external.groups, "splunk"),
+        option(true, list(external.groups, "dbs")),
+        option(false, external.groups))
+
+    # logins will be a new trait added to the cert.
+    logins: >
+      list(
+        "ubuntu",
+
+        regexp.replace(external.username, "-", "_"),
+
+        match(external.email,
+          option("nic@goteleport.com", "root")),
+
+        match(regexp.replace(external.email, "^.*@goteleport.com$", "teleporter"),
+          option("teleporter", email.local(external.email)),
+          default_option("contractor")))
+```
+
+### Predicate Helper Functions
+
+A set of helper functions to be used in the predicate expressions will be
+included.
+Part of the reasoning for using the
+[predicate language](https://github.com/vulcand/predicate)
+is so that it will be extensible.
+As the need arises, we can always add new helper function.
+
+The predicate parser is based on the `Go` language parser, so `Go` keywords
+which may be better names (such as `select`, `default`) must unfortunately be
+avoided.
+
+#### `list(...items)`
```

Comment:
nit: flatten describes the function better in my opinion

Reply:
True, changed to `flatten`. Also added basic `list` helper which I think looks better when there is nothing to flatten, and could catch mistakes if accidentally passing a list.